text/0114-cpu-threshold-read-pool.md

```diff
@@ -0,0 +1,89 @@
+
+## Current implementation
+TiKV monitors unified read pool statistics every 10 seconds(READ_POOL_THREAD_CHECK_DURATION). Based on these statistics, it adjusts the number of yatp threads by scaling up/down by one per epoch. The yatp framework handles scaling internally. It can create upto the maximum number of threads specified by the readpool.max-threads configuration but will not activate more threads than are currently needed. The size of the running task queue adjusts automatically based on the active thread count.
+
+TiKV scale up the thread count when all the following conditions are met:
+- The current number of threads is less than the maximum allowed.
+- Adding an additional thread does not overload the CPU.
+- All read pool threads are busy (threads’ busy time is at least 80%). It includes both blocking and running time.
+- There are more than 3 tasks waiting in the running queue.
+
+TiKV scale down thread count when all the following conditions are met:
+- The current thread count exceeds the configured limit.
+- The average thread utilization drops below the low water mark (70%).
+- There are fewer than 3 running tasks in the scheduling queue.
+
+## Detailed design
+TiKV will monitor cpu stats of the read pool threads to scale up/down the number of threads.
+TiKV scale down the thread count when all the following conditions are met:
+- CPU usage of unified read pool is more than the readpool.unified.cpu-threshold
+- It can bring the total number of threads less than readpool.unified.min-thread-count. 
+
+TiKV scale up the thread count when all the following conditions are met
+- CPU usage of unified read pool is less than the readpool.unified.cpu-threshold
+- Total number of threads less than readpool.unified.min-thread-count
+
+### Code changes
+
+Code would be added to calculate the CPU usage of unified read pool threads
+
+```
+get_unified_read_pool_cpu() {
+    // get all the read pool tids
+    // thread::full_thread_stat() to get the cpu usage
+}
+```
+
+adjust pool size
+
+```
+READ_POOL_THREAD_CHECK_DURATION 10 sec
+adjust_pool_size() {
+    if !self.auto_adjust {
+        return
+    }
+
+    leeway = .01; // 10%
+    if unifiedReadPoolCPU > (leeway + 1) * CPUTHreshold  {
+        target_threads = max(cur_threads * (CPUThreshold * num_cores) / unifiedReadPoolCPU, CPUThreshold * num_cores);
+        // scale down incrementally by target_threads;
+        return
+    }
+
+    if unifiedReadPoolCPU < (leeway - 1) * CPUTHreshold && activeThreads < minThreadCnt {
+        // scale up incrementally
+        return
+    }
+
+    if (unifiedReadPoolCPU > (leeway - 1) * CPUTHreshold && activeThreads >= minThreadCnt {
+        // existing code to scale up/down decided by busy thread and running tasks
+    }
```

Comment:
```suggestion
    if unifiedReadPoolCPU < CPUThreshold {
        // existing code to scale up/down decided by busy thread and running tasks
    }
```

How about leaving the scale up to the existing code, just handling the scale down logic in this optimization?

Reply:
LGTM